Баженова Анна

bazhenova opened the pull request.

bazhenova committed:
Merge remote-tracking branch 'urfu-2015/master'

bazhenova committed:
Решение задачи

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,180 @@
 
-    // Ваша невероятная магия здесь
+function isNameCorrect(name) {
+    var reg = /^[a-z0-9а-я\s]+$/i;
```
Имена могут содержать точку «Иванов И.И.» или «Поклонский-горский». А может и символы из других алфавитов «â». Поэтому проверять регуляркой смысла нет. Достаточно проверить на существование и на длину, если необходимо.

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,180 @@
+
+function isPhoneCorrect(phone) {
+    var reg = /^(\+?\d{1,2}\s*)?(\(\d{3}\)|\d{3})+\s*\d{3}(\s*|\-)?\d(\s*|\-)?\d{3}$/;
+    return reg.test(phone);
```
Прежде чем запихивать в регулярку, можно проверить что phone вообще есть и это строка.

Аналогично для двух других проверок.

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,180 @@
 
-};
+function isEmailCorrect(email) {
+    var reg = /^[-(\w|а-я).]+@([a-z0-9а-я]+[-а-яa-z0-9]+\.)+[а-яa-z0-9]{2,4}$/i;
```
Вот такой валидный email не проходит `ёжик@на.ёлке.ru`  Всё дело в том, что `ё` не входит в диапазон `а-я`

gogoleff replied:
Спецификация адресов электронной почты очень обширна и проверять на диапазоны символов бессмысленно. А наша цель избавиться от спама и не дать бабуленьке ввести ФИО вместо почты.

Поэтому достаточно проверять по формату `{символы без @}@{символы без @, но хотя бы с  одной точкой}`

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,180 @@
 
+function getRecord(record) {
+    var info = '';
+    var keys = Object.keys(record);
```
В задаче обозначены конретные поля и в решении ты проверяешь только эти конкретные поля.
А раз они заранее известны и их всего три, зачем их обходить циклом.

Для решения достаточно просто их соединить.

Есть такой хороший принцип [YAGNI](https://ru.wikipedia.org/wiki/YAGNI), старайся следовать ему для решения любых задач, как продуктовых, так и учебных. Он поможет значительно сэкономить время и проще проводить рефакторинг, легче покрывать тестами.

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,180 @@
    Функция удаления записи в телефонной книге.
 */
 module.exports.remove = function remove(query) {
+    var records = module.exports.find(query);
```
:+1: за использование find

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,180 @@
    На вход может прийти что угодно, будьте осторожны.
 */
 module.exports.add = function add(name, phone, email) {
+    if (isNameCorrect(name) && isEmailCorrect(email) && isPhoneCorrect(phone)) {
```
:+1:

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,180 @@
+            var keys = Object.keys(record);
+            for (var j = 0; j < keys.length; j++) {
+                var lowerRecord = record[keys[j]].toLowerCase();
+                if (lowerRecord.indexOf(query) != -1) {
```
Всегда используем строгое сравнение!

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,180 @@
+    }
+    newPhone += '+';
+    newPhone = newPhone.split('').reverse().join('');
+    return newPhone + '  ';
```
Форматировать телефон проще регуляркой и методами строк. Например так:

``` js
phone = phone.replace(/[^\d]+/g, ''); // Вырезаем всё кроме цифр

var formatted = '+' + (phone.slice(0, -10) || 7) + ' ' + // Берём символы с нулевого до 10 с конца (код города, если их нет ставим 7)
phone.slice(-10).replace(/(\d{3})(\d{3})(\d{2})(\d{2})/, '($1) $2-$3-$4'); // Берём 10 символов с конца.
```

bazhenova replied:
Ого, действительно круто)

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,180 @@
-    // - Разбираете записи из `data`
-    // - Добавляете каждую запись в книгу
+    var count = 0;
+    var records = data.split('\r\n');
```
В разных операционных системах строки могут оканчиваться и просто на `\n`, лучше предусмотреть все варианты:

``` js
var records = data.split('(\r\n|\n)'); // | – означает  «или»
```

bazhenova replied:
Даже не подумала об этом.

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,180 @@
+        newRecord.push(newName);
+    } else {
+        newRecord.push(record.name);
+    }
```
А если имя будет больше 13? Это число должно определяться как максимальное из возможных.

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,180 @@
+    var records = data.split('\r\n');
+    for (var i in records) {
+        var record = records[i].split(';');
+        if (module.exports.add(record[0], record[1], record[2])) {
```
За использование уже готовой функции :+1:

gogoleff commented:
@bazhenova Посмотрел, жду исправлений

**Дедлайн по доработкам продлили ещё на неделю!**

bazhenova committed:
Исправления

bazhenova commented:
@gogoleff Исправила все замечания.

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,202 @@
-var phoneBook; // Здесь вы храните записи как хотите
-
+var phoneBook = []; // Здесь вы храните записи как хотите -> хочу лист словарей ^^
+var max_name_length = 5; // для вывода
```
Константы именуем заглавными буквами

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,202 @@
-
+var phoneBook = []; // Здесь вы храните записи как хотите -> хочу лист словарей ^^
+var max_name_length = 5; // для вывода
+var max_email_length = 7;
```
Переменные именуем в формате camelCase – `maxEmailLength`

bazhenova replied:
Ой, забыла, простите)

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,202 @@
 
-    // Ваша невероятная магия здесь
+function isNameCorrect(name) {
+    return (name !== 'undefined' && name.length > 0);
```
Если параметр не пришёл, то он имеет значение и тип `undefined`, а не является строкой 'undefined'.

``` js
undefined !== 'undefined'
```

Необходимо проверять именно на `undefined`, а ещё, что не пустая строка, а ещё, что не `null`.
Подумай, как можно объеднить все эти проверки одним условием.

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,202 @@
+        return false;
+    }
+    var reg = /^[^@]+@[^@]+(\.[^@]+)+$/i;
+    return reg.test(email);
```
Можно объединить логические операторы в один:

``` js
return typeof email === 'string' &&  email.length > 0 && /^[^@]+@[^@]+(\.[^@]+)+$/i.test(email);
```

Если какое-то условие отвалится, то следующие проверяться не будут

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,202 @@
+        return false;
+    }
+    var reg = /^(\+?\d{1,2}\s*)?(\(\d{3}\)|\d{3})+\s*\d{3}(\s*|\-)?\d(\s*|\-)?\d{3}$/;
+    return reg.test(phone);
```
Можно объединить логические операторы в один

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,202 @@
 
+function getRecord(record) {
+    var info = record.name + ', ' + record.phone + ', ' + record.email;
+    return info;
```
Переменная здесь не нужна, лучше сразу:

``` js
return record.name + ', ' + record.phone + ', ' + record.email;
```

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,202 @@
-    // - Добавляете каждую запись в книгу
+    var count = 0;
+    var records = data.split('(\r\n|\n)');
+    for (var i in records) {
```
Для работы с массивами используем конструкцию `for (var i = 0; i < records.length; i++ )`
Она существенно быстрее и гарантирует порядок итерирования.

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,202 @@
+    phone = phone.replace(/[^\d]+/g, '');
+    var formatted = '+' + (phone.slice(0, -10) || 7) + ' ' +
+    phone.slice(-10).replace(/(\d{3})(\d{3})(\d{2})(\d{2})/, '($1) $2-$3-$4');
+    return formatted + '  ';
```
Если пробелы хранишь для таблички, то лучше эту логику перенести туда, а здесь не использовать пробелы

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,202 @@
+        for (var i = 0; i < max_email_length - record.email.length; i++) {
+            newEmail += ' ';
+        }
+        newRecord.push(newEmail);
```
Здесь видно повторение кода, его можно вынести в метод `padString(string, length)`

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,202 @@
+        }
+        if (name.length > max_name_length) {
+            max_name_length = name.length;
+        }
```
:+1:

gogoleff commented:
@bazhenova Ещё немного пожеланий

bazhenova committed:
Еще исправления

bazhenova commented:
@gogoleff Исправила замечания. При повторной проверке оказалось, что вот такая конструкция var records = data.split('(\r\n|\n)'); в задаче со звездочкой не работает. Вернула в изначальное состояние var records = data.split('\r\n');

gogoleff commented:
@bazhenova Да, я опечатался, там необходимо передавать не строку, а регулярное выражение:

``` js
var records = data.split(/\r\n|\n/);
```

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,191 @@
 
-    // Ваша невероятная магия здесь
+function isNameCorrect(name) {
+    return (name || false) && typeof name === 'string';
```
Можно просто  `name && typeof name === 'string'`

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,191 @@
-};
+function isEmailCorrect(email) {
+    var reg = /^[^@]+@[^@]+(\.[^@]+)+$/i;
+    return typeof email === 'string' && email.length > 0 && reg.test(email);
```
Не хватает ещё `email &&  typeof email === 'string'` и в телефоне так же

gogoleff commented on phoneBook.js:
```diff
@@ -1,52 +1,191 @@
-    // Ваша удивительная магия здесь
-
+    var records = [];
+    if (typeof query !== 'undefined') {
```
Мне кажется мы ожидаем только строку, и если она не пришла, то подменяем пустой:

``` js
query = typeof query === 'string' ? query : '';
```

gogoleff commented:
В целом всё уже очень читаемо, просто и более-менее оптимально, зачёт :)

gogoleff commented:
:star: За решение дополнительного задания
